File: src-tauri/src/events.rs
use serde::Serialize;

use crate::models::ItemStatus;

#[derive(Debug, Clone, Serialize)]
pub struct TransferProgressPayload {
    pub transfer_id: String,
    pub bytes_done: u64,
    pub bytes_total: u64,
}

#[derive(Debug, Clone, Serialize)]
pub struct TransferStatusChangedPayload {
    pub transfer_id: String,
    pub file_name: String,
    pub status: ItemStatus,
}

#[derive(Debug, Clone, Serialize)]
pub struct ConnectionStatusPayload {
    pub connected: bool,
    pub peer_address: Option<String>,
    /// "inbound" for a peer that connected to this machine's listener (this side can
    /// only receive), "outbound" for a connection this machine initiated with
    /// `connect` (this side can send). Kept distinct so the UI doesn't offer a "Send
    /// a file" button on the listening/receiving side, where it would always fail.
    pub role: &'static str,
}

/// Abstracts event emission away from the concrete `tauri::AppHandle` type so the
/// engine (and its tests) never need to reference Tauri/WebView2 types directly.
/// The only implementation used by the real app lives in `tauri_event_sink.rs`;
/// headless contexts (tests) just pass `None`.
pub trait EventSink: Send + Sync {
    fn emit_transfer_progress(&self, payload: TransferProgressPayload);
    fn emit_transfer_status_changed(&self, payload: TransferStatusChangedPayload);
    fn emit_connection_status(&self, payload: ConnectionStatusPayload);
}

File: src/lib/api/events.ts
import { listen } from "@tauri-apps/api/event";
import { connectionStatus } from "../stores/connection";
import { transferProgress, transferLog } from "../stores/transfers";

export interface TransferProgressPayload {
  transfer_id: string;
  bytes_done: number;
  bytes_total: number;
}

export interface TransferStatusChangedPayload {
  transfer_id: string;
  file_name: string;
  status: { status: string; kind?: string; message?: string };
}

export interface ConnectionStatusPayload {
  connected: boolean;
  peer_address: string | null;
  role: "inbound" | "outbound";
}

export async function startEventListeners() {
  await listen<TransferProgressPayload>("transfer:progress", (event) => {
    transferProgress.update((map) => ({ ...map, [event.payload.transfer_id]: event.payload }));
  });

  await listen<TransferStatusChangedPayload>("transfer:status-changed", (event) => {
    const { transfer_id, file_name, status } = event.payload;
    const label =
      status.status === "error"
        ? `Failed (${file_name}): ${status.message ?? status.kind ?? "unknown error"}`
        : status.status === "done"
          ? `Completed: ${file_name}`
          : `${file_name}: ${status.status}`;
    transferLog.update((log) => [{ transferId: transfer_id, label, at: Date.now() }, ...log].slice(0, 20));
  });

  await listen<ConnectionStatusPayload>("connection:status", (event) => {
    connectionStatus.set(event.payload);
  });
}
